src/commands.rs

```diff
@@ -662,6 +645,88 @@ pub fn json_set<M: Manager>(manager: M, ctx: &Context, args: Vec<RedisString>) -
     }
 }
 
+///
+/// JSON.MSET <key> <path> <json> [[<key> <path> <json>]...]
+///
+pub fn json_mset<M: Manager>(manager: M, ctx: &Context, args: Vec<RedisString>) -> RedisResult {
+    let mut args = args.into_iter().skip(1);
+
+    if args.len() < 3 {
+        return Err(RedisError::WrongArity);
+    }
+
+    // Collect all the actions from the args (redis_key, update_info, value)
+    let mut actions = Vec::new();
+    while let Ok(key) = args.next_arg() {
+        let mut redis_key = manager.open_key_write(ctx, key)?;
+
+        // Verify the key is a JSON type
+        let key_value = redis_key.get_value()?;
+
+        // Verify the path is valid and get all the update info
+        let path = Path::new(args.next_str()?);
+        let update_info = if path.get_path() == JSON_ROOT_PATH {
+            None
+        } else if let Some(value) = key_value {
+            Some(KeyValue::new(value).find_paths(path.get_path(), &SetOptions::None)?)
+        } else {
+            return Err(RedisError::Str(
+                "ERR new objects must be created at the root",
+            ));
+        };
+
+        // Parse the input and validate it's valid JSON
+        let value_str = args.next_str()?;
+        let value = manager.from_str(value_str, Format::JSON, true)?;
+
+        actions.push((redis_key, update_info, value));
+    }
+
+    actions
+        .drain(..)
+        .for_each(|(mut redis_key, update_info, value)| {
+            let updated = if let Some(update_info) = update_info {
+                !update_info.is_empty() && apply_updates::<M>(&mut redis_key, value, update_info)
+            } else {
+                // In case it is a root path
+                redis_key.set_value(Vec::new(), value).unwrap()
+            };
+            if updated {
+                redis_key.apply_changes(ctx, "json.mset").unwrap();
+            }
+        });
+    REDIS_OK
+}
+
+fn apply_updates<M: Manager>(
```

Comment:
Just to make sure, you just extracted the code above into a function right?

Reply:
one change, the `for ui in update_info` now doesn't stop on the first error

Reply:
@gkorland
> the for ui in update_info now doesn't stop on the first error

Consider testing that a specific change, which cannot be done, does not prevent `apply_changes` for previously done changes, e.g., once first change is done, second cannot be done, but third should be done as well.
```
json.set a $ '{"x": {"y":[10,20], "z":[30,40]}}'
json.set b $ '{"x": 60}'

json.mset a $.x '{}' a '$.x.z[1]' '50' b $.x '70'
```

Reply:
Maybe also on the same key
```
json.mset a $.x '{}' a '$.x.z[1]' '50' a $.u '70'
```

Reply:
Since it is also fixed for `json_set` - now it may notify on a similar scenario with multi-value (using wildcards, etc.) , while it did not notify before - could affect RediSearch results for example, but this would be a progression and is not considered a breaking change.
Agree?

Reply:
An example for the problem with a single `json.set` (using RediSearch as a subscriber for keyspace notifications)
```
127.0.0.1:6379> json.set k1 $ '{"x":{"x":{"y":[4,5,6]}}}'
OK
127.0.0.1:6379> ft.create idx on json schema '$..y[*]' as y numeric
OK
127.0.0.1:6379> ft.search idx '@y:[4 5]'
1) (integer) 1
2) "k1"
3) 1) "$"
   2) "{\"x\":{\"x\":{\"y\":[4,5,6]}}}"
```
The following `json.set` was failing in the middle and not notifying on changes, so the index was not being updated causing search to return a stale result:
```
127.0.0.1:6379> json.set k1 '$..x' '1'
(nil)
127.0.0.1:6379> json.get k1 $
"[{\"x\":1}]"
127.0.0.1:6379> ft.search idx '@y:[4 5]'
1) (integer) 1
2) "k1"
3) 1) "$"
   2) "{\"x\":1}"
```
With this PR it should be fixed ☝🏼 (search should return an empty result)
And once RediSearch/RediSearch#3507 is merged, it should also work with `json.mset`, e.g., 
```
json.mset k1 $ '{"x":{"x":{"y":[4,5,6]}}}' k1 '$..x' '1'
```
And then `ft.search idx '@y:[4 5]'` should return an empty result
And with
```
json.mset k1 $ '{"x":{"x":{"y":[4,5,6]}}}' k1 '$..x' '1' k1 '$.y' '[4, 5, 6]'
```
The same search should return k1.

<br>
I suppose such tests should be added to RediSearch/RediSearch#3507

Reply:
@gkorland Keyspace notification can also be tested in this PR (without RediSearch) as in https://github.com/RedisJSON/RedisJSON/blob/0b898bef64912f7d5837c2a2495969c92f2c2c77/tests/pytest/test_keyspace_notifications.py#L14